dipy/align/cpd.py

```diff
@@ -184,48 +184,60 @@ class DeformableRegistration:
         Number of eigenvectors to use in lowrank calculation.
     """
 
-    def __init__(self, X, Y, sigma2=None, alpha=None, beta=None,
-                 low_rank=False, num_eig=100, max_iterations=None,
-                 tolerance=None,  w=None, *args, **kwargs):
+    def __init__(
+        self,
+        X,
+        Y,
+        sigma2=None,
+        alpha=None,
+        beta=None,
+        low_rank=False,
+        num_eig=100,
+        max_iterations=None,
+        tolerance=None,
+        w=None,
+        *args,
+        **kwargs,
+    ):
```

Comment:
can we change this rule for function ? This take too much space for no reason...

What do you think?  What is your opinion @arokem also ? 

I know that @Garyfallidis  prefer too maximize the line-length instead of 1 argument per line.

Reply:
> can we change this rule for function ?

I'd say that this is determined by the line length: I am not aware of a setting that tells `ruff`: "add as many arguments as possible on the same line until the line length, and break the line only then". If you happen to find it, please let me know and I'll adapt it if Ariel and Elef are against the current style.

I know it is hard to read sometimes, and it requires effort to get used to it. This is one of prices to pay for the automation/consistency.

Reply:
@skoudoro @Garyfallidis @arokem after rebasing the failures in `code-format` are again the expected ones (https://github.com/dipy/dipy/pull/3216#issuecomment-2106265738) so if CI's come green this should be considered for merging. Rebasing a PR with 4 commits changing 33K line is hard, also because changes need to be done to the first commit, and at times I forget. Also, note that I had to make >400 manual changes because `ruff` was finding statements that did not comply with PEP8 in the strict sense. So I'd be grateful if you could afford some time to consider this PR.

Reply:
ok, I am going ahead and merge. but many formatting are still strange to me like my comment above and many array formatting. we might need use `fmt off and on` for the array.

Many new PRs will be needed to finalize/adapt this work

Reply:
also, the correction for `SimilarityMetric` deriving from `abc.ABCMeta` seems incorrect following: https://docs.astral.sh/ruff/rules/meta-class-abc-meta/

Reply:
Concerning import sorting, I got my answer in the links below:
https://github.com/astral-sh/ruff/issues/2600
https://docs.astral.sh/ruff/settings/#lint_isort_split-on-trailing-comma

Reply:
> we might need use fmt off and on for the array.

Seems not very practical to me, and leads back to manual formatting/inconsistency. But feel free to have a go.

I will have a look at https://github.com/dipy/dipy/pull/3216#discussion_r1604243313 and https://github.com/dipy/dipy/pull/3216#discussion_r1604250350.

Reply:
OK, had a look at the `ABC` inheritance. Unless other changes are made, a simple `class C(ABC):` will not do the job: you will then get
```
dipy/align/metrics.py:97:5: B027 `SimilarityMetric.use_static_image_dynamics` is an empty method in an abstract base class, but has no abstract decorator
dipy/align/metrics.py:136:5: B027 `SimilarityMetric.use_moving_image_dynamics` is an empty method in an abstract base class, but has no abstract decorator
```

and if you add the decorator to those functions, the CI's start failing, so either the methods need to be reimplemented in child classes with a simple `pass` or I do not see a way to inherit from `ABC`. @skoudoro feel free to have a stab at this.